pkg/codegen/schema/schema.go

```diff
@@ -1527,6 +1614,152 @@ type FunctionSpec struct {
 	IsOverlay bool `json:"isOverlay,omitempty" yaml:"isOverlay,omitempty"`
 }
 
+func emptyObject(data RawMessage) (bool, error) {
+	var objectData *map[string]RawMessage
+	if err := json.Unmarshal(data, &objectData); err != nil {
+		return false, err
+	}
+
+	if objectData == nil {
+		return true, nil
+	}
+
+	return len(*objectData) == 0, nil
+}
+
+func unmarshalFunctionSpec(funcSpec *FunctionSpec, data map[string]RawMessage) error {
+	if description, ok := data["description"]; ok {
+		if err := json.Unmarshal(description, &funcSpec.Description); err != nil {
+			return err
+		}
+	}
+
+	if inputs, ok := data["inputs"]; ok {
+		if err := json.Unmarshal(inputs, &funcSpec.Inputs); err != nil {
+			return err
+		}
+	}
+
+	if multiArgumentInputs, ok := data["multiArgumentInputs"]; ok {
+		if err := json.Unmarshal(multiArgumentInputs, &funcSpec.MultiArgumentInputs); err != nil {
+			return err
+		}
+	}
+
+	if returnType, ok := data["outputs"]; ok {
+		isEmpty, err := emptyObject(returnType)
+		if err != nil {
+			return err
+		}
+
+		if !isEmpty {
+			if err := json.Unmarshal(returnType, &funcSpec.ReturnType); err != nil {
+				return err
+			}
+
+			if funcSpec.ReturnType != nil && funcSpec.ReturnType.ObjectTypeSpec != nil {
+				// derive outputs from the return type
+				funcSpec.Outputs = funcSpec.ReturnType.ObjectTypeSpec
+			}
+		} else {
+			funcSpec.ReturnType = nil
+		}
+	}
+
+	if deprecationMessage, ok := data["deprecationMessage"]; ok {
+		if err := json.Unmarshal(deprecationMessage, &funcSpec.DeprecationMessage); err != nil {
+			return err
+		}
+	}
+
+	if language, ok := data["language"]; ok {
+		if err := json.Unmarshal(language, &funcSpec.Language); err != nil {
+			return err
+		}
+	}
+
+	if isOverlay, ok := data["isOverlay"]; ok {
+		if err := json.Unmarshal(isOverlay, &funcSpec.IsOverlay); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// UnmarshalJSON is custom unmarshalling logic for FunctionSpec so that we can derive Outputs from ReturnType
+// which otherwise isn't possible when both are retrieved from the same JSON field
+func (funcSpec *FunctionSpec) UnmarshalJSON(inputJSON []byte) error {
+	var data map[string]RawMessage
+	if err := json.Unmarshal(inputJSON, &data); err != nil {
+		return err
+	}
+	return unmarshalFunctionSpec(funcSpec, data)
+}
+
+// UnmarshalYAML is custom unmarshalling logic for FunctionSpec so that we can derive Outputs from ReturnType
+// which otherwise isn't possible when both are retrieved from the same JSON field
+func (funcSpec *FunctionSpec) UnmarshalYAML(node *yaml.Node) error {
+	var data map[string]RawMessage
+	if err := node.Decode(&data); err != nil {
+		return err
+	}
+	return unmarshalFunctionSpec(funcSpec, data)
+}
+
+func (funcSpec FunctionSpec) marshalFunctionSpec() map[string]interface{} {
```

Comment:
Raise an error if both `Outputs` and `ReturnType` are specified.

Reply:
Done ✅